### PR TITLE
[BXC-4035] Creates a ColorBarSegmentationDataset class and masking functions

### PR DESCRIPTION
**Pull request by wintere**
Creates a ColorBarSegmentationDataset subclass designed for image segmentation rather than classification. Also implements transformation functions to create a binary [0,1] mask representing regions of images with color bars. To be followed by a segmentation_workflow_service.py to train and evaluate a model with the new image, target mappings that computes cross-entropy loss as a weighted pixel-by-pixel difference between predicted color bar regions and ground truth masks.

Rather than return a mage, binary label tuple, the segmentation dataset class returns an image, mask tuple. Masks are initially computed as binary two dimensional arrays of size [H, W] where H, W are the height and width of the original image, then resized using a similar transformation function to the original image to fit into resnet. 

The correctness of masked regions computed has been tested by visual comparison of transformed masks to transformed images (using img.show() on converted tensors) in addition to the unit tests provided.

**Input Image**
image

**Input Tensor After Normalization and Resizing**
image

**Mask Target Tensor After Resizing**
image

### CODE REVIEW

**Review by bbpennel — CHANGES_REQUESTED** [review]
I think this looks good over all from reading, just a few questions and comments.

You mentioned that you removed the code you used to create images for visual comparison. It would be helpful to retain that, perhaps as a utility method that we can use when we need to inspect outcomes (either while coding or when evaluating the results of a trained model).

I'm trying to think through how we would use the labels (from annotations or the eventual model) for cropping. The current setup is capturing the exact coordinates of the color bar annotation rectangle, which is generally 0-2% from the edge of image, so there are often gaps. The "centerCrop" of the image also introduces additional gaps between the color bar region and the edge of the image, as in the example included in the PR description. So we would need to be able to reverse the centerCrop (I assume this is where capturing `image_dimensions` comes into play), and then capture the non-color bar region as a single rectangle by some means.

File: src/datasets/color_bar_dataset.py [hunk 1]
```diff
@@ -14,6 +12,7 @@ def __init__(self, config, image_paths, split = 'train'):
     self.labels = []
     self.split = split
     self.image_paths = image_paths
+    self.image_dimensions = []
```
[the remark this answers is not in the document]
Comment (wintere) [reply src/datasets/color_bar_dataset.py:15]: In this case, the list didn't populate properly if I added self.image_dimensions = [] to the SegmentationDataset constructor instead.

File: src/datasets/color_bar_segmentation_dataset.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  
+  # Must be overriden from parent class
+  def __getitem__(self, index):
+    image_data = load_for_resnet(self.image_paths[index], self.config.max_dimension)
+    label_mask = load_mask_for_resnet(self.labels[index], self.config.max_dimension)
+    label_mask = label_mask.clamp(0,1)
+    return image_data, label_mask
+
+  # Loads annotation data into self.labels in the same order they paths are listed in image_paths
+  def load_labels(self, path_to_labels):
```
[the remark this answers is not in the document]
Comment (wintere) [reply src/datasets/color_bar_segmentation_dataset.py:26]: Fixed.

File: src/datasets/color_bar_segmentation_dataset.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+     w, h = None, None
+     with Image.open(image_path) as img:
+      w, h = img.width, img.height
+     self.image_dimensions.append((w, h))
+     # Populate label masks
+     image_labels = path_to_labels[str(image_path)]
+     mask = zeros((h, w), dtype=uint8)
+     for label in image_labels:
+       if 'color_bar' in label['rectanglelabels']:
+        width, height = (label['width']), (label['height'])
```
[the remark this answers is not in the document]
Comment (wintere) [reply src/datasets/color_bar_segmentation_dataset.py]: Fixed, this was an artifact of when I converted from the relative measurements from float to integer earlier in the process.

**Comment by wintere** [comment]
OK, I've added an image/mask tensor displaying helper function and corrected the formatting errors.  The visualize_tensor command as run in the unit tests displays as expected on Windows, but I haven't tested on another OS.

**Comment by bbpennel** [comment]
The visualize_tensor method seems to work well on macOS. Maybe keep the lines in the test for getting them, but comment them out so it doesn't pop open images every time the test runs?

**Review by bbpennel — APPROVED** [review]
